### PR TITLE
Update network with firewall

### PR DESCRIPTION
**Pull request by MykolaSolyanko**

### CODE REVIEW

File: launcher/launcher.go [hunk 1]
```diff
@@ -675,9 +703,71 @@ instancesLoop:
...
+
 	return errStatus
 }
 
+func (launcher *Launcher) prepareNetworkForInstances(onlyExposedPorts bool) (errStatus []cloudprotocol.InstanceStatus) {
+	for _, node := range launcher.nodes {
+		for i, instance := range node.currentRunRequest.instances {
+			serviceInfo, err := launcher.imageProvider.GetServiceInfo(instance.ServiceID)
+			if err != nil {
+				log.WithField("serviceID", instance.ServiceID).Errorf("Can't get service info: %v", err)
```
Comment (al1img) [inline launcher/launcher.go:722]: empty line after log

Comment (MykolaSolyanko) [reply launcher/launcher.go:722]: Done

File: networkmanager/networkmanager.go [hunk 2]
```diff
@@ -21,52 +21,85 @@ package networkmanager
...
 	NetworkID string
 }
 
+// FirewallRule represents firewall rule.
+type FirewallRule struct {
+	Protocol string
+	Port     string
+}
+
+type InstanceNetworkInfo struct {
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply networkmanager/networkmanager.go:91]: Done

File: networkmanager/networkmanager.go [hunk 3]
```diff
@@ -21,52 +21,85 @@ package networkmanager
...
+	aostypes.InstanceIdent
+	NetworkInfo
+	Rules []FirewallRule
+}
+
+type NetworkParameters struct {
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply networkmanager/networkmanager.go:97]: Done

File: networkmanager/networkmanager.go [hunk 4]
```diff
@@ -172,20 +238,34 @@ func (manager *NetworkManager) RestartDNSServer() error {
...
+			params.Hosts, fmt.Sprintf("%d.%s.%s", instanceIdent.Instance, instanceIdent.SubjectID, instanceIdent.ServiceID))
 
 		if instanceIdent.Instance == 0 {
-			hosts = append(hosts, fmt.Sprintf("%s.%s", instanceIdent.SubjectID, instanceIdent.ServiceID))
+			params.Hosts = append(params.Hosts, fmt.Sprintf("%s.%s", instanceIdent.SubjectID, instanceIdent.ServiceID))
 		}
 	}
 
+	defer func() {
+		if err == nil && len(params.AllowConnections) > 0 {
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply networkmanager/networkmanager.go]: Agree, nice suggestion. Done

File: networkmanager/networkmanager.go [hunk 5]
```diff
@@ -172,20 +239,34 @@ func (manager *NetworkManager) RestartDNSServer() error {
...
+			}
+
+			networkParameters.FirewallRules = firewallRules
+		}
+	}()
+
 	networkParameters, found := manager.getNetworkParametersToCache(networkID, instanceIdent)
 	if found {
-		if err := manager.dns.addHosts(hosts, networkParameters.IP); err != nil {
+		if err := manager.dns.addHosts(params.Hosts, networkParameters.IP); err != nil {
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply networkmanager/networkmanager.go]: Done

**Review by al1img — APPROVED** [review]

**Pull request merged**